hail/python/hailtop/batch/batch.py

```diff
@@ -149,6 +151,58 @@ def __init__(self,
 
         self._cancel_after_n_failures = cancel_after_n_failures
 
+        self._python_functions: Dict[int, Tuple[Callable, Optional[str]]] = {}
+
+    def _get_python_function(self, function_id: int) -> Callable:
+        function, _ = self._python_functions.get(function_id, (None, None))
+        if function is None:
+            raise ValueError(
+                f"No Python function registered with id: {function_id}."
+            )
+        return function
+
+    def _get_python_function_file_name(self, function_id: int) -> str:
+        _, file_name = self._python_functions.get(function_id, (None, None))
+        if file_name is None:
+            raise ValueError(
+                f"No file found for Python function registered with id: {function_id}."
+            )
+        return file_name
+
+    def _register_python_function(self, function: Callable) -> int:
+        function_id = id(function)
+        try:
+            self._get_python_function(function_id)
+        except ValueError:
+            self._python_functions[function_id] = (function, None)
+        return function_id
+
+    async def _serialize_python_to_input_file(
+        self, path: str, file_id: int, code: Any, dry_run: bool = False
+    ) -> str:
+        pipe = BytesIO()
+        dill.dump(code, pipe, recurse=True)
+        pipe.seek(0)
+
+        code_path = f"{path}/code{file_id}.p"
```

Comment:
It would be nice for debugging if the functions were stored more "separately" from the args. It's all blobs in the end but it would be nice to have something like `{path}/functions/` and `{path}/args/`